### ISSUE TITLE
When reading emojis, count replicated emojis rather than reading each individually

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#8499]
This has been suggested by several users on various forums recently:

Often on social media, people post text with repeated emojis in it, for instance, tweets on Twitter or posts on Facebook.

NVDA currently reads each emoji individually, which can be very long.  Where emojis are repeated, it would be efficient for the reader to have NVDA report each emoji once, with a count for multiple instances.  I don't think I can include emojis in here, but for instance:

<Heart><Heart><Heart>I <Heart> <Chocolate><Chocolate><Chocolate><Heart><Heart><Heart>
could be read as:
<Heart>x3 I <Heart> <Chocolate>x3 <Heart>x3

**Comment by leonardder** [comment]
We already have this for punctuation I believe, so we might be able to build upon that system?

**Comment by derekriemer** [comment]
tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji tears of joy emoji

**Comment by jcsteh** [comment]
1. If we supported emoji in NVDA's builtin symbols dictionary (#6523), we'd get this for free; it'd be handled just like any other repeated character. Right now, we rely on speech synthesisers to speak emoji for us, so that isn't possible.
2. Alternatively, if we plan to continue relying on speech synths for emoji support, I guess we could add a specific rule in characterProcessing to match repeated emoji. Emoji have specific Unicode ranges, so this should be possible.

**Comment by jcsteh** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [Jul 31, 2018, 1:25 PM GMT+10](https://github.com/nvaccess/nvda/issues/8499#issuecomment-409082167 "2018-07-31T03:25:16Z - Replied by Github Reply Comments"):
> 1. If we supported emoji in NVDA's builtin symbols dictionary (#6523), we'd get this for free; it'd be handled just like any other repeated character.

Nope, I'm wrong. The repeated character rule only works for characters which consume a single UTF-16 code point. Emoji are 32 bit Unicode characters and are thus 2 UTF-16 characters. That concern alone should be eliminated by a move to Python 3, since then these would be a single code point. However, if you add a skin tone modifier, things get even messier, since now you're dealing with multiple Unicode code points regardless of encoding.

**Comment by jcsteh** [comment]
It might be possible to change the repeated character rule to handle characters with multiple code points by using a regexp or `(a|b|c)` instead of a regexp character set `[abc]`. Not quite sure why I didn't do this in the first place; might have been worried about performance.

Anyway, the question of whether we plan to continue relying on synths for emoji still needs to be answered before we can determine what to do here.

**Comment by fernando-jose-silva** [comment]
Sorry to get into a subject with a little technical basis.
It would be interesting that nvda could speak the emugis independently of the synthesizer, because there are enough interesting synthesizers that do not support the reading of emugis, as the nuance vocaliser esprecive , and the addon of synthesizer of the nv acces SpeechPlayer in Espeak

**Comment by Adriani90** [comment]
@leonardder this might be of interest for your work on this as well.

**Comment by leonardder** [comment]
Normal emoji without skin tone modifiers have been implicitly fixed with Python 3. I'm closing this for now, as I have the feeling that emoji with skin tone modifiers are special cases for which a new issue fits better.
